### ISSUE TITLE
Buffered Message Sending Fails after broker restart in background

### ISSUE THREAD

**Issue by mohitkhanna** [htgc/fluent-plugin-kafka#10]
Hi,

We have one of our FluentD setup boxes streaming into one of the brokers (the only one in our test cluster). I noticed that after restarting the broker service (different server) that the plugin was connected to, further sending of back-log'd/buffered messages fail with the below error. The issue gets corrected after restarting Fluentd service.

Looks like the Poseidon connection does not reconnect with Kafka broker - after it looses connection upon broker restart. Any help with this?

2015-01-06 12:31:07 +0000 [warn]: temporarily failed to flush the buffer. next_retry=2015-01-06 12:33:54 +0000 error_class="RuntimeError" error="Failed to send all messages: [#<Poseidon::Message:0x007ff01d1617e0 @struct=#<struct Poseidon::Protocol::MessageWithOffsetStruct offset=0, message=#<struct Poseidon::Protocol::MessageStruct magic_type=0, attributes=0, key=nil, value=\"{\\\"tstamp\\\":\\\"2015-01-06T12:27:59+00:00\\\",\\\"ver\\\":\\\"v1\\\",\\\"host\\\":......... some more data ......\\\"}\">>, @topic=\"zmi-weblogs\">, #<Poseidon::Message:0x007ff01d1616a0 @struct=#<struct Poseidon::Protocol::MessageWithOffsetStruct offset=0, message=#<struct Poseidon::Protocol::MessageStruct magic_type=0, attributes=0, key=nil, value=\"{\\\"tstamp\\\":\\\"2015-01-06T12:28:01+00:00\\\",\\\"ver\\\":\\\"v1\\\",\\\"host\\\":\\\"....... some more data .....\\\"}\">>, @topic=\"zmi-weblogs\">] remaining" plugin_id="object:d57dc0"
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/poseidon-0.0.5/lib/poseidon/sync_producer.rb:66:in `send_messages'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/poseidon-0.0.5/lib/poseidon/producer.rb:163:in `send_messages'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:112:in `block (2 levels) in publish'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:109:in `each'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:109:in `each_with_index'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:109:in `block in publish'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:108:in `each'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:108:in `publish'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluent-plugin-kafka-0.0.8/lib/fluent/plugin/out_kafka_buffered.rb:104:in `write'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluentd-0.10.57/lib/fluent/buffer.rb:300:in `write_chunk'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluentd-0.10.57/lib/fluent/buffer.rb:280:in `pop'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluentd-0.10.57/lib/fluent/output.rb:311:in `try_flush'
  2015-01-06 12:31:07 +0000 [warn]: /usr/lib64/fluent/ruby/lib/ruby/gems/1.9.1/gems/fluentd-0.10.57/lib/fluent/output.rb:132:in `run'

**Comment by htgc** [comment]
I confirmed it can resend after reconnected connection with following log.
```
2015-01-18 23:14:01 +0900 [warn]: retry succeeded. plugin_id="object:3fa3f02d1148"
```
Perhaps, I couldn't build a same environment.(I tried stop kafka-server while td-agent is working, after exception occurred, I restart kafka-server)

Please let me know how to build this situation if I make misunderstanding.
Thanks,

**Comment by mohitkhanna** [comment]
@htgc I can consistently reproduce it on my setup. I will see if I have concrete steps to reproduce it and mail them to you.

BTW... Just to add on to my observation: I can see the error happening after the max retries have been reached. So after I let the data go into the buffer file, I bring down the broker and let the agent fail more than max retries.

After that, this error starts occurring and data is never retransmitted. Lets for time being assume that that's the right behaviour, but:

1. Why repeated errors about "All messages not sent"? Looks like it's trying to attempt again and again on the same data.
2. What happens to the new data that comes from that point onwards? Will that atleast be sent? I have to again confirm that. I don't think new data is being sent unless i restart the agent.

For now - i have put a retry limit of 150 to take care of major downtime - but that's just a workaround.

**Comment by htgc** [comment]
Sorry for being super late.
I missed your notice below.
> after restarting the broker service (different server)

This plugin configures broker info only start fluentd, so it couldn't reconnect another broker server.
I'll fix this issue by next release.

Sorry for inconvenience.

**Comment by raviprakashshahi** [comment]
Hi Mohit,

Did you manage to re-push the logs in the buffer file? I am facing the same problem and had to move the buffer files for td-agent to start working again.

Logs in buffer files contain the tag-name and some special character also, something like "��TagName�X�����message� �" [logs]

Is there any plugin which can reprocess these logs.

**Comment by mohitkhanna** [comment]
Ravi,

It's been sometime so I cannot recall exact steps that we had to do. But the logs got pushed out to Kafka after we corrected some topic Leader election issue and restarting fluentd agent.

Besides we had the raw source data to help us recover in case that step had not worked.

Not aware of any plugin that could have done that forced.

Best reach out to the author of the plugin for that. Also Fluentd authors are very active on Slack - suggest you join that and ask this question.

Search for fluentd + Slack and you'd know how to onboard on that forum.

Regards,
Mohit
